# Computer setup

## Docker

# Tuesday

## ID intro with Aga
Indit 

PCV product
NNPP

NGSA

Hugues - Pronounced Ugh
Sumana

CEPI pan betacorona


BCE
Humoral response - B cell
T cell and b cell team. Separate teams.
Aga is on the B cell team too.
Which is part of the data science team.

You will do engineering onboarding too.

Tokyo team. Technical bioinformaticians and researchers.

- [ ] Post an introduction on slack with some background on your

We'll find a task in Jira.
They have a backlog because of no engineer work.
Hanna knows the product well.

Onboarding checklist for HR.

1 on 1 with everyone encouraged.
Ask people to reach out for a 1 on 1 in your intro message.
Then reach out to the rest.
Hey i am this, just joined do you have time to ... talk a bit?


IT onboarding.

Hanna and Aga in heidelberg are away next week.

Alex is in all meetings, Matheus too. 
Alex is in the translational team with Matheus, team of chief scientific officer.
Historically: RTD - matheus and -alex.
Per Brattås - jonied all meetings.

Two more scientefic developers joining.
Vaccine design.
Software development.

Team meetings alrnate between check ins and discussion meetings.
We alternate in writing the agenda.
Quick scale after the meeting.

We fill out actions points in the documents

Pair programming.

You can use your budget for training and books.
You can add it to your 
Conference budget.

A meeting with the Tool.

- [x] Add your name to the check in
  
- Ask Marius for the task. Extract scores from the pipeline.
- Improve testing 
- Try to break the pipeline. See if the error message handles it well or if you want to improve it
- run indit on hertzner servers

You can break the servers, that are for ID

Youness pipeline was breaking stuff, figure out what the problem was and see it.
Pin the dependency -> maybe you can bump some other package.

- [x] Write intro message

## Team meeting

First letters of our name and surname
Youness and Per work 50%.
We need a new acronym. ALOHAG
Matheus is dropping out of the team meetings.

Scarda: Funding agency from Japan.
Has a mlaria, dengue, CCHF and ebola project.


Hanna: Clustering pipeline output. Seeing if it covers the input sequences and 
any HLAs?
Can we predict what HLAs will present different parts of a virus? A model for this maybe?

Who is Daikey?
What is AIDD, alternative funding agency from Scarda.

Finding memory of people that are infected, or to validate that new peptides work in this way.

RNN binding, CPP, HOPP.
Extract binding scores from the pipeline.
60GB of results, okay to keep on the server.

Scarda servers are used for scarda analysis only.

rsync files between servers to keep them in sync, recursively in the morning.
or get it in a bucket. How?

Lazy frames loads only the columns you need?
Polars. Learn Polars?

No inferiority to the BioNTech vaccine.

Two hires in Heidelberg, will work remote.

Noelia presented my tool at the Rosetta conference.

NLE has many hires. They do bioinformatics at NLE?
Multiple hires for MD, protein design, at NLE, Kaidra manages this?

What is a dark antigen?

Run with flags on the peptides. Flags are amino acids surrounding the 
sequences.

Binding and processing scores, binding to what? degested peptides to T cell presented on HLA?
Processing data. Processing by ... 

Pipeline can be run on peptides or on sliding window of full proteins.
All possible 9 and 10 meers on 5k proteins some up to 1k amino acids.
So the peptides can be very short.

Per: Peptide HLA coverage thing in PR review.
Code looks okay, Marius will run it on test data.
Hotspot epitope score.

Company acronyms add an acronym without a definition and tag a person.

- [ ] Set up 1v1 with Per and Youness this week
- [ ] Set up a 1v1 with your desk mate Sumana
- [x] Akillesh?
- [ ] Kaider?
- [x] Schedule INDIT intro with Marius
    - I want to run the pipeline myself first


I want to get a learning habit going each day. 
What and when?
Double whammy, part in work hours part at home. Evening routine?
At work libraries and things, like polars, etc
At home: biology, immunology.

## Running indit
Set up cursor for oncoimmunity w gogle account. I get a 14 day pro trial!

I need to install python.
sqlite, and docker.
I think I should be able to run something locally no? Some light tests, a simple
test case fex?
Try to run something now?

Uses pip tools instead of poetry.
pip-compile generates requriements.txt with exact hashses

Input fasta sequences with proteins.


## Indit intro w Marius


8-9mers hla class 1, hla a and hla b
12

156 a and B hlsa

data science team will tell you
Mass spectrometry data.

 9 mers and hla pair singl escore
 score is not a probability but will be later.

9 and 10 mers

including flanks

Curate the scores, we don' need 7-8gb of intermediate scores per protein.

we can curate this to a more useful set, not write everything to disk fex.

Germnline calling 1 caller
consensus calling for tumor dna, minmum read requriement

trna too, some variants will be in geneses that are turned of.

No expression in the ID pipeline.

cancer mutation load differs 
maybe 200 variants for a patient with scorers,
you pik top 20 scores
make a vaccine iwht 5 of them with linkers

neojunktions  for the linkers

**ID**
heatmap of AP scores.
what is the max score of all the mers that this amino acid is part of.
It is a heat mapt over the sequence an all the hlas

explainability - Peyman works with this.
Talk to them about them to understand which scores contribute the most to the AP score,
and which are stable.

Hotspots which part of the protein has the highest immunogenci potential

Bins of 25 mers from the heatmap, tried this for a while. Do any bins have higher potential?
You can binarize the scores too.
Bins had problems with the hard edges.

- [x] Ask for hotspot identification pdf
- [ ] Create notes summarizing the talk
Sum of all the hla scores for an amino acid.
Finding mountains with sums of scores.
Peak regions output for your input set of proteins


Haplotype for HLAs , a and b types.

Digital twin,

Popultation distributions

Step 0: Pre processin
Finding proteins, often public data. You need to do litterature review and learn the avg sequence length etc.
Sounds really nice to do with claude projects!
Step 1: AP scores
Step 2: Hotspot prediction

A good viewer for spreadsheets on linux?

tsv is better than csv.

orthopox family - small pox.
What hot spots overlap between variants of orthopox viruses.
Goal: Find a vaccine that is protective for all orthopox viruses

One vaccine can have 2-3 10mers
Each peptide can only be 100 amino acids.

15mer so you can 
mRNA vaccines can have thousands of amino acids.

Some times we use longer pieces like 100+ amino acids segments, or you
can snip up a lot 

Sequencing price of mer length.

The 

Group them by amino acid composition.
Fingerprints betwen hotspots between species.
This seems like a sequence alignment score really.
Multiple sequence alightmens

Beta corona 275 species.
what hotspots will cover all them.
How robust are the hotspots to mutation in the species?
Will you vaccine be 

DFS - distance for self models.
Only at the very end.
Compare all 9mers you have to all 9mers in the human proteome.
Takes a long time.

None of the 9mers had a match for hepatitis.
Vital tissues.
This is all T-cell stuff.
Youaniss is leading the B-cell team.

Known B cell epitopes.
Find which regions we can have a B cell response too.

Conformational epitopes. If I mutate at this spot we can give 
protection against all of these specief for this virus fex.


How can we ground it in something, what is the AP score?
Some of the HLAs are biased.


## Reviewing PRs by trym
I am surprised at how small the diffs are
20 lines refactors.
PR feedback about keeping a docstring or not
nit about reducing code duplication

This will be an adjustment. Very tiny changes.
I can do that though if that is what people like
here. But I am curious what people think about
speed?
Maybe this stuff is so critical because it is
used with real patients and there are documentation
requirements?
I can codify all these comments and do my own review with cursor
before I open a PR if I want to deliver with this quality and detail
myself.

What happened to Cursor?
I can't run commands from spacemacs in here anymore?
Restart fixed it.

# Wednesday
- [x] Set up 1 on 1 with youness
- [ ] 1 on 1 with Per

I could do 1 more this week? Pinalee?
Or kaider?
You can set up talks for next week as well now?

Ghazal


##  Haakon 1v1
TG4050 valideres i mennesker

Both cancer and infectious disease.
Structure is important.

RSA relative accessible surface area
score for alle aminosyrer.
SS3
SS9
Dihedral angels

Lager modeller som predikerer ting med proteinene, 
scores osv som brukes som input til 
Generer nye epitoper med permutasjoner.
Vegard og Matheus

Autoencoder for RSA


Hjerte bilder ultralyd, PHD
GE wingmed
Billig
Ultralyd kan måuk flow med doppler


DNT turleder, trening
https://www.dnt.no/dnt-der-du-er/dnt-oslo-og-omegn/aktiviteter/



## Reviewing Jira 

Hanna makes a detailed plan of her week, tracks estimated and actual times 
for what she does. She accounts for the full week.
https://neconcoimmunity.atlassian.net/jira/software/c/projects/ID/issues/ID-289?jql=project%20%3D%20%22ID%22%20ORDER%20BY%20created%20DESC
Something to consider.

I wonder if I could do the same? Do you want to?

Here is the bug Aga wanted you to start on
https://neconcoimmunity.atlassian.net/jira/software/c/projects/ID/issues?jql=project%20%3D%20%22ID%22%20ORDER%20BY%20created%20DESC

## Run indit 
11:31 - 11:45

- book a hertzner machine ✅
cepi9
- ssh to cepi9 ❌
Blocked, no shh key until tomorrow.

12:20 -
- read docs on how to run
- run tests locally


## Data science intro
What wsa NIP again?

MHC class 1
---Antigen presentation---
Expression, translation
Processing
    - Digestion
    - ER translocation (endoplasmic retriculum)
HLA binding
Immunogenicity (presentation). Likelyhood that a T-cell receptor recognizes this.
-----

Cytosol, endosome?
Transporation is done by TAP proteins.

MHC class 2

Personalization is because people have different HLA-A, B and C genes.
class 1 length: 8-12 amino acids

NAQ - neo antigen quest. NLA developed it. Separate pipeline which predicts
binding for class II ~15 mers.

We need 30 vaccin elements.
We only get 15 with our pipelen. We would run NAQ and fill it up with 
class II elements (up to 15 mers).

One model per allele? As in the allele that determines your which HLA versions you have.

NLA has churn, and worse documentation.

**Binding models**:
Input: peptide sequence, hla allele
Output binding affinity: Binding affinity 0-1, converted to IC50.
4 models: noi-binding
HOPP: k-NN graph model. patented
CPC: patented.
- both developed by people that left the company, same for resert
LSTMBind: Ioannis?
Used as an ensamble model. Used as features in XGboost will learn which of them are most important.
They don't judge the model, they are all used as input. But how do you train this ensemble model?
Ah the models are pre-trained, and you're training your ensamble thing.

Public training data. IEDB etc.
https://www.iedb.org/ -- anki

Class I
Proteasome chops into random lengths.
Tap transports everything.
If things bind to MHC I it will be presented.

Class II
Cells can engulf viruses, canerous cells, etc and chop them up with endosome, proteases
cleave stuff up and loads it onto class MHC II.
Seems like these are important for infectious disease no?

**Processing models**
input: 8-12mers, + optional 0-3 flanks.

Mass spec can tell oyu if something was present on the cell surface, and htat it was bound to a mhc.
Mass spec?
We only get positive instances from the biological data.
We need to generate negative peptides. That were not processed. Didn't understand how.

12 support vector machines - build by consultatns when the company was started
12 feed forward neural networks,
- 3 negative generation strategies w 3 training datasets
- 2 peptide masking stategies
- 2 amino acid encodings, VHSE and one hot. -- anki VHSE, could use a pLM embedding
Encoding amino acids individually is a bit odd? They get properties because they are a 
chain that has 3D properties. protein language models would learn this.

You could use a protein language model embedding to encode an amino acid?

**AP models**
repo: antigen model builder
Input
source transcript?
Expression, how much of a protein is there in het cell

Output:
ranking of how lkiely we are to present
(future: probability of being preseted)

Expression levels, disabled in ID pipeline.

Mass spectrometry, we can know that the peptides are bound to a mhc on the surface of a cell line.
Also do RNA seq. -- ANKI mass spectrometry
They have some proprietary data too for different cell lines, what presents on the surface of the cell.
last two weeks they looked at it: Normalized per HLA so they can't be compared between HLAs.
HLA-A and HLA-B are comparable usually, HLA-C has much less data.

**DIstance from self: Immune presentation models**
Is it enough to be foreign? I.e not in the patient proteome.

For a known immunogenic peptide we look for the background(the peptides that are most similar but not immunogenic)
then they learn a distance metric?

Missing data to train good models for distance to self.


IEDB says if a peptide is immunogenic or not?

**dataset pre-processing**
Pulled from IEDB
Int - short for int int - short for intelligent integration

Scotty: Runs on gcp, get the

1:1000 ratio of peptides from the proteome that actually are shown on the 
surface of proteins.
That is why they generate 1000 negatives per positive.

xAI dashboards what is important in our pipeline. Peyman does this.


A new model is a new subclass of an XGboost model with the name 
has info on what features are used, x - includes expression, r means ... 

NLE served AIDD before they bought NOI. NLE projects are PhD and master projects
often published.

Bio-renderer for rendering the notes.
Could a remarkable be nice to have? TO make notes on this pdf fex?
And to read the bio books?
What about the sunlight computer tablet? To read and take notes.

**NIP**
NEC immuno profiler.
NNPP is the whole thing NIP + whatever the above thing is called.

NIP identifies neo antigens, 27mer regions enriched around a neo antigen.
And do the vaccine design.

Susanne gives the intro to NIP.

Look at the architecture diagram.


**cleanup**
- remove cepi9 booking
https://neconcoimmunity.atlassian.net/wiki/spaces/ID/pages/1153826994/Infectious+Diseases+Server+Booking

**could do**
- Read docs about the ID team https://neconcoimmunity.atlassian.net/wiki/spaces/ID/pages/1790050360/ID+onboarding
has info on work practice, indit.
- set up theme command
- set up calendar invite
- order the cell, and the immunology book, (paper pluss digital?), and a pc mouse

- set up pymol
- think about what could be automated with LLMs, boundaryml

## Ioannis
CPC model
Hopp

BCE i 5 [r alene]

Antigens
where on the protein will they bind?
given a protein

3d epitopes
interne algorimer to say if something is on the surface or if it is part of a secondary strcuture
instead of alphat fold
RSA, SS, HSE

Alphafold beats us, use it?

b cells do kill cancer when the surface porteins are mutated.

 common practice is to predict binary is this part of an epitope for a b cell

 Engineer an epitope to be protective against all 300 variants
 epitope engirneering

 BLSDM
 LSTM er et 

 started using GPUS

 Vaccine delivered 
 260 amino acids, mutated version of the spike protein

B cells vs t cels.
All infectious disease is b cells?

B cells in livmorshallskreft.

 B og T cell engineering
 da er det viktig aa finne akkurat rett epitop.

levert som antibodies
uten minne

https://neconcoimmunity.atlassian.net/wiki/spaces/BCE/pages/318409966/Immunology+Notes

Bilde av whiteboard
https://oncoimmunity.slack.com/files/U0816THHFNH/F081Q6LD9B4/img_8996.jpg
https://www.rcsb.org/3d-view/1MHP

# Thursday

To buy
- Bigger monitor for desk
- Logitech mouse
34 inch monitors for 3k roughly on dusin.
https://www.dustin.no/group/hardware/datautstyr-tilbehor/skjermer-tilbehor/skjermer?displaydiagonal=31.5,32,34,37.5&price=1000-1500,1500-2000,2000-2500,2500-3000,3000-3500,3500-4000,500-750,750-1000
https://www.dustin.no/product/5011332372/viewfinity-s50gc 34 inch 3440*1400 2400 nok

## Run INDIT

Installed python 3.11 w pyenv
make dev failed
ERROR: No matching distribution found for oiml==4.7.1

Seems like that is on gemfury

Commented out oiml from requirements-dev.txt for now.
After insalling all deps in the .venv it installs the package indit itself in editable mode.
To do this it looks for the dependcies in the pyproject.toml file.

I think this is the repo for the oiml package
https://github.com/OncoImmunity/oncoimmunity-machine-learning

Searching in a file name in github is useful
oiml  path:**/pyproject.toml

Installed oncoimmunity-machine-learning from the local files by 

```sh
source indit/venv/bin/activate
cd oncoimmunity-machine-learning
pip install -e .
```



Tasks!
- align python version in confluence and readme, 3.9 or 3.11?
- Unpin pydantic version in pyproject.toml, the issue it links is resolved now     # pydantic <2 because of https://github.com/PrefectHQ/prefect/issues/10145
    - Test prefect
- Fix update
- check if PIP_EXTRA_INDEX_URL is set on server
  - Copy the token and set PIP_EXTRA_INDEX_URL locally

## Hanna intro
class2-scorer
15 mers
20 HLAs only for class 2

mhc-bert

Use for further scoring of class 1 hotspots first,
but then run separately as a 

Picking vaccine elements for cchf
hand picked some proteins relevant for cchf
5 proteins.
we have 10-1000s of variants per protein

clusters of hotspots accross versions of the same protein
We want clusters that cover a lot of the versions.
We have 1 representative sequence that covers 91% of the seqences.
So this is probably in a well conserved region of the protein.

For this hotspot. How many hls do each peptide cover.
Can we find some really good peptides with even higher coverage in the
input sequences.
And coverage accrros many HLAs?
Coverage accross HLA means score over 0.64.

Now run class 2 scoring for the same hotspot, can we see if it hits there too.

.parquet file keeps polars dataframes

input:
-hotspot[]

splits into 14,15,16 mers
preps data format for bertmhc
spins up docker for bertmhc
runs berthmc

outputs the score for pairs of
- n-mer - 
- hla-class 2

ouput

We can't run for the full sequences because it takes a long time.
Runs on a cpu I think.
GPU it up! Less of a problem to run for a lot of sequences.

A way to have a CI run with only the production dependencies.
How could we do that? To capture things working when installed.

## Staff meeting

Judged by:
* Results
* Code of values

Tech meeting in the staff meeting.
Summary of what the teams have accomplished.
So we can have more knowledge sharing.
I like it.

**Tech meeting slides**, format is 1 slide per team, high level, quick and dirty
https://drive.google.com/drive/folders/16Cyby49dyBTkTv6lMjp6Xr-Eg0kWJwFN
https://docs.google.com/presentation/d/1peBWrQE1sBGehdzJQQ7U25bhl1V7WMdkEOBGTlLLvfg/edit#slide=id.g30086d2b9f2_3_10 useful to get up to speed on what is being worked on now.

Fail early and fail well.

50 minute and 25 minute meetings instead of 1 hours so we have a buffer between.
Telephone booths? where are they?

Mark yourself as WFH when you work from home. I don't want to type each day that I am
working from home.

Talk to each others, as meetings are starting.

All servers are booked.
Ping someone on slack if you see that a server is idle.
Hertzner server status page where is it?

Servers can be used at the same time.
SLURM solved this quite well.
Time limit for interactive jobs.

How can I check how much space is available on a server?

**activity dashbaord for heztner in Grafana**
http://lr1:3000/d/ZsYvfWUIk/machine-usage-dashboard?orgId=1

## GPUs in NOI
They create VMs in GCP compute engine.
https://neconcoimmunity.atlassian.net/wiki/spaces/COL/pages/56690108/Setting+up+a+cloud+machine+on+GCP#Deprecation-motice
Which you can assign a GPU (T4)
https://neconcoimmunity.atlassian.net/wiki/spaces/RD/pages/1805287436/How+to+set+up+a+GCP+VM+with+GPU
Billed 200$ for a month of a T4.

VM instances, several with GPUs. Several are stopped.
https://console.cloud.google.com/compute/instances?inv=1&invt=AbiEFA&project=genuine-sector-223709
BCE seems to use GPU machines.
![](2024-11-21-09-57-58.png)

**Buckets in GCP**
https://console.cloud.google.com/storage/browser?referrer=search&inv=1&invt=AbiEFA&project=genuine-sector-223709&prefix=&forceOnBucketsSortingFiltering=true
You can create and dump stuff in your own buckets



## Onboarding to engineering
The engineering onboarding might mention the grafana account?
- https://neconcoimmunity.atlassian.net/wiki/spaces/OP/pages/1268350982/Engineering+Onboarding

It mentions documens for how to work and best practice

- [ ] I want to start reading this later.
- [ ] I also want to schedule a 1 on 1 with Stian to get some more context.

Graphana password is in bitwarden
There's an item in bitwarden called "Grafana" that you should be able to see. The username and password are there

## Per
NGSA 
HLA og variant calling, RNA expression data, what are the levels

NGSA + NIP = NNPP
NNPP=NEC Neo antigen prediction ppipeline

AIDD team in japan, gets an image
RNA and DNA seq input -> vaccine candidate for a patient

Take samples and sequence- 1-2 weeks
Sequencing people upload to a gcp bucket


2 ongoing studies

TG4050 is a study
Trans gene gets neoantigen predictions from us.
They optimize the vaccine delivery stuff based on 30 fragments.
Cloned into a virus, viral vectors. developed by 
~1 patient a week

Trans gene: TCR sequence
TCR sequencing before and after, more of our T cells.
We can't know which T cells we target. So we can only show a change.
We can test what T cells respond to our vaccine elements.
Predicing that this T cell will bind this elements would be huge, can we model it.

HLA alleles er en peronsalisering 

Study results
https://www.transgene.fr/wp-content/uploads/20241108_TG4050_SITC2024_Poster.pdf

Vaxxim - aquired - second trial phase 1.
Oral delivery.
Klones inn i salmonella.
Skal reagere i tarmen.

## Lunch - GPU talk
Aga wants to GPU-ify the whole pipeline.
I can do that!

## NEC bio science kickoff

Nishihara san? From the NEC group?
NEC co-technical officer?
https://jpn.nec.com/profile/corp/executives/bio/m-nishihara.html
Corporate Executive Vice President and CTO (Executive Officer)
Joined NEC in 1985
Moved under GIUB, which he runs?

The AI drug development will grow for the future of NEC.
This is a new business area for NEC.
Market is European and North America.
Well aligned with NEC's goals. AIPP is a prominent candidate.
DIBU

https://www.nec.com/en/global/about/executives/exc-officers.html
Nishira San is one of 20 executives.
Akimura San.
NEC is an expert in AI systems for IT services. What is AI for IT services?
Alvaro Aranguren - clinical study manager at NEC Bio therapeutics.

Corona vaccine at CEPI.
![](2024-11-21-12-18-40.png)

Therapeutic vaccines, all cancer vaccines so far are for therapeutic vaccines.

TCR therapy.

Business development is the biggest goal this and next year.
We need to start generating revenue next year.

Engineered T-cell receptor vaccines.

Heinz Lubenau - from Vaxxim. Worked there for 14 years.
DS - 

CDA - Confidentiality disclosure agrement. Goal of 3 per year.
1 due dilligence initiation.

CEPI targeting 1 new grant.

13k whole genome sequences from japan. Relevant for T cell therapy somehow?

https://www.nec-bio.com/ - this shows **all the projects**

Gritstone and Nykode are both having big trouble.
Nykode - 80m $, Genentech dropped partnership
Gritstone - 3.5m $

Genentech is part of Roche group.
https://www.gene.com/medical-professionals/pipeline/#our-pipeline
They have many drugs in their pipeline.

We need to manage clinical trial results and expectation to avoid a tank.

NEC has invested in BostonGene series A.

Kaidre worked with Kitamura San Akira.

They are starting to see NEC as a credible player in AIDD.

TG4050
Vaccine after standard treatment. 
Standard treatment is surgery.
After 24 months there are no relapses in the relapse arm.

BioNTech our selection of targets seems to be better?
Targets are our vaccine elements?

Better immune response than the pfizer vaccine for more recent variants of the corona virus.
![](2024-11-21-12-44-47.png)


Design of enchancer TCRs against cancer.

ADC targeted therapy - antibody drug conjugates.

Schrodringer has a 150m$ deal with Novartis.
They do the vaccine design too.

Pharma has a remarkably low AI approach?

We can find targets really well.
But we need to tell Pharma how to design their drug too. For this we need
work on Scaffolds and on vector design for the delivery
![](2024-11-21-12-51-09.png)


**Hein Lubenau - NEC biotherapeutics**
NEC biotherapeutics wants to get clinical trials ready by 2025.

Basket trial for metastic cancer in Lithuania.

They measure:
Tumor growht - can we shrink tumors with treatment.

NECVAX-NEO1
n=15, n=20, n=8 trials.

60 questoins for approval with the German arm of the trial.

Ethics committe - we needed to do another round of informed conscent.

AGC Biologics - what was biologics?

They produce the personalized vaccine in 3 weeks? From the arrival of the plasmids I think?
Is that fast? HE says it is unparalleled.

There are way to many abbreviations in this business. If we're gonna levrage data scientist, 
machine learning engineers, software engineers we need to keep the language simple.

**NOI Saverio**
The company chart was useful.
![](2024-11-21-13-15-10.png)
Moritz, Erlend and Trym are software engineers - learn from them!
4 people from parental leave.

CEPI nanoball, 1.2mEUR in funding.

We can make a blastingly fast ID pipeline all on GPUs. With a dashboard for interpretability.
Click?

BD = business development efforts.
How can we increase the interest here?
It is the responsability of everybody.
Jasinta & Christian are the key responsible people.

Sanofi & openai deals for patient recruitment.
InstaDeep aquired by BioNTech.
It will be their computational arm.

I feel very motivated.

**AIDD org structure Minori Saitoh**
AIDD has several business developers.
![](2024-11-21-13-28-17.png)

Akira Kitamura

**Actions**:
Draft a message to Saverio to explain
what we could get out of me going to the AIDD conference
in Barcelona.
Connect with Ferruz again and with Patrick.
Potential hires, brief them on what NOI does, if they
could be interesting to us.
Bryan fex? After he publishes.


## Class2 scorer PR review and test run

Had to install gcloud cli and authenticate docker with gcloud
After update, the following will be written to your Docker config file located at [/home/aksell/.docker/config.json]:
gcloud auth configure-docker.

```
I had a go at running everything in the README. The instructions were clear and I could get most every thing working without trouble, really nice!

The one thing that confused me was that I didn't set up git-lfs initially to pull the model weights. Maybe add a line to the readme in the BertMHC section about pulling the model weighs from git lfs before
building the image.

See my comment about the change you saw you needed for the `packages` list in pyproject.toml.
```



# Friday
## Fix error in workflow tests
https://oncoimmunity.slack.com/archives/C043JQ13EN9/p1731923281323569

Had to set prefect logging path to run the tests
```shell
prefect config set PREFECT_LOGGING_SETTINGS_PATH="$(pwd)/configs/logging.yaml"
```

change hostname? It is very long now. Changed from admubuntu-ThinkPad-X1-Carbon-Gen-12 -> carbon-x1

https://github.com/docker/docker-py/issues/3256 requests issue
Fix released in docker 7.1 https://github.com/docker/docker-py/pull/3257

Prefect 2.18 does not support docker 7.
Check what prefect versions are there without bumping major version.
>>> pip index versions prefect | grep "2\."

See deps of a specific version! Very useful
>>> pip download prefect==2.20.14 --no-deps --no-binary :all: && tar xf prefect-2.20.14.tar.gz && cat prefect-2.20.14/requirements.txt

Prefect 2.20 (latest 2.x release) support docker >= 4.0, < 8.0.
If you think we could bump the prefect major version I'm happy to have a go at that too.
I wanted to avoid bumping the prefect major version here, since prefect seems quite core and a new major version might need some rewrites/changes
that I don't know if I should do now before having the setup to run the full pipeline on server and some familiarity with how we validate
the pipeline.

After prefect and docker bump tests seem to run slower?
Failed with actual failing tests now:
```shell
'Run validate_input script on another file with invalid filename' done.

workflow_tests/test_indit_pce.yaml ...............................                                                                                                                                                                                                                                                                                        [ 22%]
workflow_tests/test_junction_scorer.yaml .....                                                                                                                                                                                                                                                                                                            [ 26%]
workflow_tests/test_pipeline.yaml ........F........................F.F..F.F.............F.....F...F.F..........                                                                                                                                                                                                                                           [ 83%]
workflow_tests/test_scripts.yaml .......................                                                                                                                                                                                                                                                                                                  [100%]One or more tests failed. Keeping temporary directories and logs. Use '--kwd' or '--keep-workflow-wd' to disable this behaviour.


=========================================================================================================================================================================== FAILURES ============================================================================================================================================================================
_________________________________________________________________________________________________________________________________________________________________________ test session __________________________________________________________________________________________________________________________________________________________________________
Observed md5sum '017de3fc1b62a438990b51b6442f901a' not equal to expected md5sum 'c1bef6cb1272c1ea20b3b768174c5c9b' for file '/tmp/pytest_workflow_5z6op6n5/Run_indit_pipeline_on_multiple_input_files/Results/hotspot-epitopes-scorer/results_10mers/table_all_epitopes_per_representative_sequence_per_hla.tsv'
_________________________________________________________________________________________________________________________________________________________________________ test session __________________________________________________________________________________________________________________________________________________________________________
'0_52,E,Human_coronavirus_OC43,1,9,MFMADAYLA,   ,DTV,HLA-A*01:01,17944.83,12385.0,,30081.0,,,,0.3069,False,,,xgbfp_chrp13,' was not found in /tmp/pytest_workflow_5z6op6n5/Run_indit_pipeline_on_multiple_input_files/Results/ap_predictions/cepi21_all_predictions.csv while it should be there.
_________________________________________________________________________________________________________________________________________________________________________ test session __________________________________________________________________________________________________________________________________________________________________________
Observed md5sum '7ed1ea8f67270cdaaf52dbf274c1559e' not equal to expected md5sum 'b065c6707b72923b55a5482ac81c7d53' for file '/tmp/pytest_workflow_5z6op6n5/Run_indit_pipeline_on_multiple_input_files/Results/ap_predictions/cepi21_all_predictions.csv'
_________________________________________________________________________________________________________________________________________________________________________ test session __________________________________________________________________________________________________________________________________________________________________________
'0_180,NS5,Bovine_coronavirus,1,9,MDIWKPEIK,   ,YLR,HLA-A*01:01,23726.877,14907.0,,114274.0,,,,0.3018,False,,,xgbfp_chrp13,' was not found in /tmp/pytest_workflow_5z6op6n5/Run_indit_pipeline_on_multiple_input_files/Results/ap_predictions/cepi22_all_predictions.csv while it should be there.
_________________________________________________________________________________________________________________________________________________________________________ test session __________________________________________________________________________________________________________________________________________________________________________
Observed md5sum 'f688b877f7d962a7461277e0f045ebd4' not equal to expected md5sum '0e7cb93ed97d2ebbbca8db393f92cb30' for file '/tmp/pytest_workflow_5z6op6n5/Run_indit_pipeline_on_multiple_input_files/Results/ap_predictions/cepi22_all_predictions.csv'
_________________________________________________________________________________________________________________________________________________________________________ test session __________________________________________________________________________________________________________________________________________________________________________
Observed md5sum '695715f5cf47dfaf9dc4ddb9bd9bc617' not equal to expected md5sum 'ce81680e6ce303bcc7d162b5be139f3c' for file '/tmp/pytest_workflow_5z6op6n5/Run_indit_pipeline_on_multiple_input_files/Results/ap_predictions/intermediate_predictions/cepi22_peptides_ap.csv'
_________________________________________________________________________________________________________________________________________________________________________ test session __________________________________________________________________________________________________________________________________________________________________________
Observed md5sum 'bb1b6c29e4037c61d36f72a311b2519c' not equal to expected md5sum '9201e2bd100462a927f202fe2e5b2aee' for file '/tmp/pytest_workflow_5z6op6n5/Run_indit_pipeline_on_multiple_input_files/Results/ap_predictions/intermediate_predictions/cepi21_peptides_ap.csv'
_________________________________________________________________________________________________________________________________________________________________________ test session __________________________________________________________________________________________________________________________________________________________________________
Observed md5sum '74bcffa2b7a707e0211cae9627b3e883' not equal to expected md5sum '0a9a304d98c94996a41fef9e79fe2088' for file '/tmp/pytest_workflow_5z6op6n5/Run_indit_pipeline_on_multiple_input_files/Results/ap_predictions/intermediate_predictions/cepi22_rnn.csv'
_________________________________________________________________________________________________________________________________________________________________________ test session __________________________________________________________________________________________________________________________________________________________________________
Observed md5sum 'bc9c68df2b42a12ddc96b1527b315a67' not equal to expected md5sum '343a6286708baace8bf3cb783a6e0a44' for file '/tmp/pytest_workflow_5z6op6n5/Run_indit_pipeline_on_multiple_input_files/Results/ap_predictions/intermediate_predictions/cepi21_rnn.csv'
==================================================================================================================================================================== short test summary info ====================================================================================================================================================================
FAILED workflow_tests/test_pipeline.yaml::Run indit pipeline on multiple input files::Results/hotspot-epitopes-scorer/results_10mers/table_all_epitopes_per_representative_sequence_per_hla.tsv::md5sum
FAILED workflow_tests/test_pipeline.yaml::Run indit pipeline on multiple input files::Results/ap_predictions/cepi21_all_predictions.csv::content::contains '0_52,E,Human_coronavirus_OC43,1,9,MFMADAYLA,   ,DTV,HLA-A*01:01,17944.83,12385.0,,30081.0,,,,0.3069,False,,,xgbfp_chrp13,'
FAILED workflow_tests/test_pipeline.yaml::Run indit pipeline on multiple input files::Results/ap_predictions/cepi21_all_predictions.csv::md5sum
FAILED workflow_tests/test_pipeline.yaml::Run indit pipeline on multiple input files::Results/ap_predictions/cepi22_all_predictions.csv::content::contains '0_180,NS5,Bovine_coronavirus,1,9,MDIWKPEIK,   ,YLR,HLA-A*01:01,23726.877,14907.0,,114274.0,,,,0.3018,False,,,xgbfp_chrp13,'
FAILED workflow_tests/test_pipeline.yaml::Run indit pipeline on multiple input files::Results/ap_predictions/cepi22_all_predictions.csv::md5sum
FAILED workflow_tests/test_pipeline.yaml::Run indit pipeline on multiple input files::Results/ap_predictions/intermediate_predictions/cepi22_peptides_ap.csv::md5sum
FAILED workflow_tests/test_pipeline.yaml::Run indit pipeline on multiple input files::Results/ap_predictions/intermediate_predictions/cepi21_peptides_ap.csv::md5sum
FAILED workflow_tests/test_pipeline.yaml::Run indit pipeline on multiple input files::Results/ap_predictions/intermediate_predictions/cepi22_rnn.csv::md5sum
FAILED workflow_tests/test_pipeline.yaml::Run indit pipeline on multiple input files::Results/ap_predictions/intermediate_predictions/cepi21_rnn.csv::md5sum
========================================================================================================================================================== 9 failed, 127 passed in 3860.24s (1:04:20) ===========================================================================================================================================================
cursor.AppImage: *** [Makefile:37: workflow-test] Error 1
(venv) ➜  indit git:(main) ✗ 
```
I want to know if it is caused by the docker or prefect dump or not.
I will check if I get theses errors if I just checkout the old requests version.

Ergonomic orgasm in magit: b b on a commit in hisory to check it out withou typing any hash!

Okay so the error I got after reverting to the old repo version seemes to be due to prefect db migration not being reversable.
After `rm -rf ~/.prefect` I will se what happens.
So prefect has a db that needs to be migrated, good to know.

Running this in the terminal is annoying, another reason to use direnv?
```shell
prefect config set PREFECT_LOGGING_SETTINGS_PATH="$(pwd)/configs/logging.yaml"
```

Commit edit msg is written with vscode.
What if you try to uninstall vscode? Can you write the commit messages in cursor then?

Bump docker 6.0.1->7.1.0 and prefect 2.18.1->2.20.14

The bump of requests described in https://neconcoimmunity.atlassian.net/browse/ID-292
was causing trouble. The issue is described in this docker-py issue
https://github.com/docker/docker-py/issues/3256 which links the fix 
released in docker 7.1 https://github.com/docker/docker-py/pull/3257

Prefect 2.18 doesn't support docker 7 but the latest 2.x release of
prefect 2.20.14 does. Therefore prefect is bumped too.
I wanted to avoid bumping prefect all the way to 3.x because
1. It isn't neccesary to resolve the issue
2. It might require code changes and more validation. I didn't look
at the changelog for prefect 3.x so I don't know how major of a change it is.

That said, if upgrading to prefect 3.x is a goal, I'd be happy to have a go 
at that too, now that it is being bumped anyway.
